target_test.go

```diff
@@ -0,0 +1,394 @@
+	encryptedQuery, context, err := target.odohKeyPair.Config.Contents.EncryptQuery(obliviousQuery)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	request, err := http.NewRequest(http.MethodPost, queryEndpoint, bytes.NewReader(encryptedQuery.Marshal()))
+	if err != nil {
+		t.Fatal(err)
+	}
+	request.Header.Add("Content-Type", odohMessageContentType)
+
+	rr := httptest.NewRecorder()
+	handler.ServeHTTP(rr, request)
+
+	if status := rr.Result().StatusCode; status != http.StatusOK {
+		t.Fatal(fmt.Errorf("Result did not yield %d, got %d instead", http.StatusOK, status))
+	}
+	if rr.Result().Header.Get("Content-Type") != odohMessageContentType {
+		t.Fatal("Invalid content type response")
+	}
+
+	responseBody, err := ioutil.ReadAll(rr.Result().Body)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	odohQueryResponse, err := odoh.UnmarshalDNSMessage(responseBody)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	response, err := context.OpenAnswer(odohQueryResponse)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if !bytes.Equal(response, r.queryResponseMap[q]) {
+		t.Fatal(fmt.Errorf("Incorrect response received. Got %v, expected %v", response, r.queryResponseMap[q]))
+	}
+}
+
+func TestQueryHandlerODoHWithInvalidKey(t *testing.T) {
+	r := createLocalResolver(t)
+	target := createTarget(t, r)
+
+	handler := http.HandlerFunc(target.targetQueryHandler)
+
+	differentKeyPair := createKeyPair(t)
+	q := r.queries[0]
+	obliviousQuery := odoh.CreateObliviousDNSQuery([]byte(q), 0)
+	encryptedQuery, _, err := differentKeyPair.Config.Contents.EncryptQuery(obliviousQuery)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	request, err := http.NewRequest(http.MethodPost, queryEndpoint, bytes.NewReader(encryptedQuery.Marshal()))
+	if err != nil {
+		t.Fatal(err)
+	}
+	request.Header.Add("Content-Type", odohMessageContentType)
+
+	rr := httptest.NewRecorder()
+	handler.ServeHTTP(rr, request)
+
+	if status := rr.Result().StatusCode; status != http.StatusBadRequest {
+		t.Fatal(fmt.Errorf("Result did not yield %d, got %d instead", http.StatusBadRequest, status))
+	}
+}
+
+func TestQueryHandlerODoHWithCorruptCiphertext(t *testing.T) {
+	r := createLocalResolver(t)
+	target := createTarget(t, r)
+
+	handler := http.HandlerFunc(target.targetQueryHandler)
+
+	q := r.queries[0]
+	obliviousQuery := odoh.CreateObliviousDNSQuery([]byte(q), 0)
+	encryptedQuery, _, err := target.odohKeyPair.Config.Contents.EncryptQuery(obliviousQuery)
+	if err != nil {
+		t.Fatal(err)
+	}
+	queryBytes := encryptedQuery.Marshal()
+	queryBytes[len(queryBytes)-1] ^= 0xFF
+
+	request, err := http.NewRequest(http.MethodPost, queryEndpoint, bytes.NewReader(queryBytes))
+	if err != nil {
+		t.Fatal(err)
+	}
+	request.Header.Add("Content-Type", odohMessageContentType)
+
+	rr := httptest.NewRecorder()
+	handler.ServeHTTP(rr, request)
+
+	if status := rr.Result().StatusCode; status != http.StatusBadRequest {
+		t.Fatal(fmt.Errorf("Result did not yield %d, got %d instead", http.StatusBadRequest, status))
+	}
+}
```

Comment:
Just something small I noticed, none of the tests send multiple requests to the same target server which from an external perspective could lead to a server modifying some state and not being able to serve different requests. Could be useful to create a common test that sends a couple different bad requests and then a good request.

For instance, you can merge the invalid key, corrupt ciphertext and standard odoh request function into one where all three send requests to the same server.

Reply:
We can do this in a future change.